File: CODE/beale.f90
! n = 2
! -4.5 <= x <= 4.5
! fstar = 0
!**************************************************
!	FUNZIONE BEALE
!**************************************************

      SUBROUTINE STARTP(N,X)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(N)

      X(1)=1.0D0
      X(2)=1.0D0

      RETURN
      END
!**************************************************
!	FUNZIONE BEALE
!**************************************************
	SUBROUTINE FUNCT(N,X,F)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
        PARAMETER (M=3)
	DIMENSION X(N),Y(M)
        Y(1)=1.5D0
        Y(2)=2.25D0
        Y(3)=2.625D0
        IF(DABS(1.D0-X(2)).GT.1.D+16) GO TO 1
        IF(DABS(X(1)).GT.1.D+5) GO TO 1
        IF(DABS(X(2)).GT.1.D+5) GO TO 1
	F=(Y(1)-X(1)*(1.D0-X(2)))**2
        IF(DABS(1.D0-X(2)**2).GT.1.D+16) GO TO 1
	F=F+(Y(2)-X(1)*(1.D0-X(2)**2))**2
        IF(DABS(1.D0-X(2)**3).GT.1.D+16) GO TO 1
	F=F+(Y(3)-X(1)*(1.D0-X(2)**3))**2
	RETURN
    1   F=1.D+34
        RETURN
	END
!**************************************************
!	GRADIENTE
!**************************************************
	SUBROUTINE GRAD(N,X,G)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
        PARAMETER (M=3)
	DIMENSION X(N),Y(M),G(N)
        Y(1)=1.5D0
        Y(2)=2.25D0
        Y(3)=2.625D0
	G(1)=-2.D0*(Y(1)-X(1)*(1.D0-X(2)))*(1.D0-X(2))
	G(1)=G(1)-2.D0*(Y(2)-X(1)*(1.D0-X(2)**2))*(1.D0-X(2)**2)
	G(1)=G(1)-2.D0*(Y(3)-X(1)*(1.D0-X(2)**3))*(1.D0-X(2)**3)

	G(2)=2.D0*(Y(1)-X(1)*(1.D0-X(2)))*X(1)
	G(2)=G(2)+2.D0*(Y(2)-X(1)*(1.D0-X(2)**2))*(2.D0*X(1)*X(2))
	G(2)=G(2)+2.D0*(Y(3)-X(1)*(1.D0-X(2)**3))*(3.D0*X(1)*X(2)**2)
	RETURN
	END
!**************************************************	
!	HESSIANO
!**************************************************
	SUBROUTINE HESS(N,X,H)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
        PARAMETER (M=3)
	DIMENSION X(N),Y(M),H(N,N)
        Y(1)=1.5D0
        Y(2)=2.25D0
        Y(3)=2.625D0
	H(1,1)=2.D0*(1.D0-X(2))*(1.D0-X(2))
	H(1,1)=H(1,1)+2.D0*(1.D0-X(2)**2)*(1.D0-X(2)**2)
	H(1,1)=H(1,1)+2.D0*(1.D0-X(2)**3)*(1.D0-X(2)**3)
      
	H(2,2)=2.D0*X(1)*X(1)
	H(2,2)=H(2,2)+2.D0*(2.D0*X(1)*X(2))**2 +2.D0*(Y(2)-X(1)*(1.D0-X(2)**2))*(2.D0*X(1))
	H(2,2)=H(2,2)+2.D0*(3.D0*X(1)*X(2)**2)**2 +2.D0*(Y(3)-X(1)*(1.D0-X(2)**3))*(6.D0*X(1)*X(2))

	H(1,2)=-2.D0*(1.D0-X(2))*X(1) +2.D0*(Y(1)-X(1)*(1.D0-X(2)))
	H(1,2)=H(1,2)-2.D0*(1.D0-X(2)**2)*(2.D0*X(1)*X(2)) +2.D0*(Y(2)-X(1)*(1.D0-X(2)**2))*(2.D0*X(2))
	H(1,2)=H(1,2)-2.D0*(1.D0-X(2)**3)*(3.D0*X(1)*X(2)**2) +2.D0*(Y(3)-X(1)*(1.D0-X(2)**3))*(3.D0*X(2)**2)

	H(2,1)=H(1,2)

	RETURN
	END
!**************************************************
!	HESSIANO x DIREZIONE
!**************************************************
	SUBROUTINE HESDIR(N,X,D,HD)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
	DIMENSION X(N),HD(N),D(N),H(2,2)
      	CALL HESS(N,X,H)
        DO 2 I=1,N
      	HD(I)=0.	 
      	DO 2 J=1,N
    2  	HD(I)=HD(I)+H(I,J)*D(J)	
     	RETURN
	END

File: Python/beale.py
# n = 2
# -4.5 <= x <= 4.5
# fstar = 0
#**************************************************
#	FUNZIONE BEALE
#**************************************************

import numpy as np

def startp():
    x = np.zeros(2)
    x[0] = 1.0
    x[1] = 1.0
    return x

def funct(x):
    y = np.array([1.5, 2.25, 2.625])
    
    if abs(1.0 - x[1]) > 1.0e-16 or abs(x[0]) > 1.0e5 or abs(x[1]) > 1.0e5:
        return 1.0e34
    
    f = (y[0] - x[0] * (1.0 - x[1]))**2
    f += (y[1] - x[0] * (1.0 - x[1]**2))**2
    f += (y[2] - x[0] * (1.0 - x[1]**3))**2
    
    return f

def grad(x):
    y = np.array([1.5, 2.25, 2.625])
    g = np.zeros(2)

    g[0] = -2.0 * (y[0] - x[0] * (1.0 - x[1])) * (1.0 - x[1])
    g[0] -= 2.0 * (y[1] - x[0] * (1.0 - x[1]**2)) * (1.0 - x[1]**2)
    g[0] -= 2.0 * (y[2] - x[0] * (1.0 - x[1]**3)) * (1.0 - x[1]**3)

    g[1] = 2.0 * (y[0] - x[0] * (1.0 - x[1])) * x[0]
    g[1] += 2.0 * (y[1] - x[0] * (1.0 - x[1]**2)) * (2.0 * x[0] * x[1])
    g[1] += 2.0 * (y[2] - x[0] * (1.0 - x[1]**3)) * (3.0 * x[0] * x[1]**2)

    return g

def hess(x):
    y = np.array([1.5, 2.25, 2.625])
    h = np.zeros((2, 2))

    h[0, 0] = 2.0 * (1.0 - x[1]) * (1.0 - x[1])
    h[0, 0] += 2.0 * (1.0 - x[1]**2) * (1.0 - x[1]**2)
    h[0, 0] += 2.0 * (1.0 - x[1]**3) * (1.0 - x[1]**3)

    h[1, 1] = 2.0 * x[0] * x[0]
    h[1, 1] += 2.0 * (2.0 * x[0] * x[1])**2 + 2.0 * (y[1] - x[0] * (1.0 - x[1]**2)) * (2.0 * x[0])
    h[1, 1] += 2.0 * (3.0 * x[0] * x[1]**2)**2 + 2.0 * (y[2] - x[0] * (1.0 - x[1]**3)) * (6.0 * x[0] * x[1])

    h[0, 1] = -2.0 * (1.0 - x[1]) * x[0] + 2.0 * (y[0] - x[0] * (1.0 - x[1]))
    h[0, 1] -= 2.0 * (1.0 - x[1]**2) * (2.0 * x[0] * x[1]) + 2.0 * (y[1] - x[0] * (1.0 - x[1]**2)) * (2.0 * x[1])
    h[0, 1] -= 2.0 * (1.0 - x[1]**3) * (3.0 * x[0] * x[1]**2) + 2.0 * (y[2] - x[0] * (1.0 - x[1]**3)) * (3.0 * x[1]**2)

    h[1, 0] = h[0, 1]

    return h

def hesdir(x, d):
    h = hess(x)
    hd = np.dot(h, d)
    return hd


if __name__ == '__main__':
    # Example Usage
    x_initial = startp()
    print("Initial x:", x_initial)

    # Call the function FUNCT
    result_funct = funct(x_initial)
    print("Result (FUNCT):", result_funct)

    # Call the function GRAD
    result_grad = grad(x_initial)
    print("Result (GRAD):", result_grad)

    # Call the function HESS
    result_hess = hess(x_initial)
    print("Result (HESS):")
    print(result_hess)

    # Example Direction vector
    direction_vector = np.array([0.1, 0.2])

    # Call the function HESDIR
    result_hesdir = hesdir(x_initial, direction_vector)
    print("Result (HESDIR):", result_hesdir)
